In [1]:
import pandas as pd

In [2]:
df = pd.read_csv('fuel_ferc1.csv')

In [3]:
df.head()

record_id  utility_id_ferc1  report_year  \
0   f1_fuel_1994_12_1_0_7                 1         1994   
1  f1_fuel_1994_12_1_0_10                 1         1994   
2   f1_fuel_1994_12_2_0_1                 2         1994   
3   f1_fuel_1994_12_2_0_7                 2         1994   
4  f1_fuel_1994_12_2_0_10                 2         1994   

       plant_name_ferc1 fuel_type_code_pudl fuel_unit  fuel_qty_burned  \
0              rockport                coal       ton        5377489.0   
1  rockport total plant                coal       ton       10486945.0   
2                gorgas                coal       ton        2978683.0   
3                 barry                coal       ton        3739484.0   
4             chickasaw                 gas       mcf          40533.0   

   fuel_mmbtu_per_unit  fuel_cost_per_unit_burned  \
0               16.590                      18.59   
1               16.592                      18.58   
2               24.130                      39.72   
3               23.950                      47.21   
4                1.000                       2.77   

   fuel_cost_per_unit_delivered  fuel_cost_per_mmbtu  
0                         18.53                1.121  
1                         18.53                1.120  
2                         38.12                1.650  
3                         45.99                1.970  
4                          2.77                2.570

In [6]:
A = [1,2,3,4,5,6]
B = [13, 21, 34]

In [9]:
print(A.extend(B))

None


In [10]:
A+B

[1, 2, 3, 4, 5, 6, [13, 21, 34], [13, 21, 34], 13, 21, 34, 13, 21, 34]

In [13]:
np.identity(3)

array([[1., 0., 0.],
       [0., 1., 0.],
       [0., 0., 1.]])

In [12]:
import numpy as np

In [14]:
np.eye(3)

array([[1., 0., 0.],
       [0., 1., 0.],
       [0., 0., 1.]])

### Which of the following fuel type code has the lowest average fuel cost per unit burned?

In [23]:
df.groupby("fuel_type_code_pudl").mean()["fuel_cost_per_unit_burned"].sort_values()

fuel_type_code_pudl
gas          13.659397
other        18.253856
waste        19.518122
coal         67.421830
oil         168.877086
nuclear    4955.157002
Name: fuel_cost_per_unit_burned, dtype: float64

### What is the standard deviation and 75th percentile of the measure of energy per unit (Fuel_mmbtu_per_unit) in two decimal places?

In [25]:
df.describe()

utility_id_ferc1   report_year  fuel_qty_burned  fuel_mmbtu_per_unit  \
count      29523.000000  29523.000000     2.952300e+04         29523.000000   
mean         118.601836   2005.806050     2.622119e+06             8.492111   
std           74.178353      7.025483     9.118004e+06            10.600220   
min            1.000000   1994.000000     1.000000e+00             0.000001   
25%           55.000000   2000.000000     1.381700e+04             1.024000   
50%          122.000000   2006.000000     2.533220e+05             5.762694   
75%          176.000000   2012.000000     1.424034e+06            17.006000   
max          514.000000   2018.000000     5.558942e+08           341.260000   

       fuel_cost_per_unit_burned  fuel_cost_per_unit_delivered  \
count               29523.000000                  2.952300e+04   
mean                  208.649031                  9.175704e+02   
std                  2854.490090                  6.877593e+04   
min                  -276.080000                 -8.749370e+02   
25%                     5.207000                  3.778500e+00   
50%                    26.000000                  1.737100e+01   
75%                    47.113000                  4.213700e+01   
max                139358.000000                  7.964521e+06   

       fuel_cost_per_mmbtu  
count         29523.000000  
mean             19.304354  
std            2091.540939  
min             -41.501000  
25%               1.940000  
50%               4.127000  
75%               7.745000  
max          359278.000000

### What is the skewness and kurtosis for the fuel quantity burned in two decimal places?

In [28]:
df['fuel_qty_burned'].skew()

15.851495469109503

In [29]:
df['fuel_qty_burned'].kurtosis()

651.3694501337732

### Which feature has missing values and what is the total number of missing value and percentage of the missing rows as a factor of the total number of rows in three decimal places? 

(Enter answer in the format: Feature: xxx, Total: xxx, Percent: xxx)

In [30]:
df.isnull().sum()

record_id                         0
utility_id_ferc1                  0
report_year                       0
plant_name_ferc1                  0
fuel_type_code_pudl               0
fuel_unit                       180
fuel_qty_burned                   0
fuel_mmbtu_per_unit               0
fuel_cost_per_unit_burned         0
fuel_cost_per_unit_delivered      0
fuel_cost_per_mmbtu               0
dtype: int64

In [32]:
df.shape

(29523, 11)

In [33]:
(180/29523)*100

0.609694136774718

- Feature with missing value is: "fuel_unit", 
- Total number of missing value is: 180,
- Percentage of missing rows is: 0.610

### The feature with missing values falls under what category? What missing value imputation technique would you use?

In [38]:
df.head()

record_id  utility_id_ferc1  report_year  \
0   f1_fuel_1994_12_1_0_7                 1         1994   
1  f1_fuel_1994_12_1_0_10                 1         1994   
2   f1_fuel_1994_12_2_0_1                 2         1994   
3   f1_fuel_1994_12_2_0_7                 2         1994   
4  f1_fuel_1994_12_2_0_10                 2         1994   

       plant_name_ferc1 fuel_type_code_pudl fuel_unit  fuel_qty_burned  \
0              rockport                coal       ton        5377489.0   
1  rockport total plant                coal       ton       10486945.0   
2                gorgas                coal       ton        2978683.0   
3                 barry                coal       ton        3739484.0   
4             chickasaw                 gas       mcf          40533.0   

   fuel_mmbtu_per_unit  fuel_cost_per_unit_burned  \
0               16.590                      18.59   
1               16.592                      18.58   
2               24.130                      39.72   
3               23.950                      47.21   
4                1.000                       2.77   

   fuel_cost_per_unit_delivered  fuel_cost_per_mmbtu  
0                         18.53                1.121  
1                         18.53                1.120  
2                         38.12                1.650  
3                         45.99                1.970  
4                          2.77                2.570

### Which of the features has the second and third lowest correlation with the Fuel Cost Per Unit Burned?

In [44]:
df.corr().sort_values("fuel_cost_per_unit_burned", ascending=True)

utility_id_ferc1  report_year  fuel_qty_burned  \
utility_id_ferc1                      1.000000     0.093323        -0.057447   
fuel_qty_burned                      -0.057447     0.012952         1.000000   
fuel_mmbtu_per_unit                  -0.066946    -0.110853        -0.080946   
fuel_cost_per_mmbtu                   0.006122     0.010261        -0.001896   
fuel_cost_per_unit_delivered         -0.016414    -0.014043        -0.003551   
report_year                           0.093323     1.000000         0.012952   
fuel_cost_per_unit_burned            -0.037863     0.013599        -0.018535   

                              fuel_mmbtu_per_unit  fuel_cost_per_unit_burned  \
utility_id_ferc1                        -0.066946                  -0.037863   
fuel_qty_burned                         -0.080946                  -0.018535   
fuel_mmbtu_per_unit                      1.000000                  -0.010034   
fuel_cost_per_mmbtu                     -0.005884                  -0.000437   
fuel_cost_per_unit_delivered            -0.009039                   0.011007   
report_year                             -0.110853                   0.013599   
fuel_cost_per_unit_burned               -0.010034                   1.000000   

                              fuel_cost_per_unit_delivered  \
utility_id_ferc1                                 -0.016414   
fuel_qty_burned                                  -0.003551   
fuel_mmbtu_per_unit                              -0.009039   
fuel_cost_per_mmbtu                              -0.000109   
fuel_cost_per_unit_delivered                      1.000000   
report_year                                      -0.014043   
fuel_cost_per_unit_burned                         0.011007   

                              fuel_cost_per_mmbtu  
utility_id_ferc1                         0.006122  
fuel_qty_burned                         -0.001896  
fuel_mmbtu_per_unit                     -0.005884  
fuel_cost_per_mmbtu                      1.000000  
fuel_cost_per_unit_delivered            -0.000109  
report_year                              0.010261  
fuel_cost_per_unit_burned               -0.000437

### For the fuel type coal, what is the percentage change in the fuel cost per unit burned in 1998 compared to 1994?

In [48]:
df[df['fuel_type_code_pudl'] == 'coal'].groupby("report_year").sum()['fuel_cost_per_unit_burned']

report_year
1994     14984.572
1995     14571.785
1996     12694.803
1997     43742.178
1998     11902.597
1999     16303.468
2000    100428.831
2001     15109.012
2002      9171.164
2003     30954.110
2004      9632.063
2005     11107.182
2006     12402.684
2007     12965.171
2008     14331.348
2009     45554.858
2010     88627.745
2011     17001.015
2012     17546.097
2013     17629.860
2014     15412.197
2015     13338.235
2016     11100.925
2017     10254.284
2018      9488.197
Name: fuel_cost_per_unit_burned, dtype: float64

### Which year has the highest average fuel cost per unit delivered?

In [53]:
df.groupby('report_year').sum().sort_values("fuel_cost_per_unit_delivered", ascending=False)

utility_id_ferc1  fuel_qty_burned  fuel_mmbtu_per_unit  \
report_year                                                           
1997                   118553     3.077973e+09         11210.909740   
1996                   118281     2.768693e+09         11726.508078   
2000                   152518     3.394100e+09         12665.204661   
2009                   144366     2.633401e+09          9394.970149   
2018                   127120     3.974124e+09          6273.315699   
2015                   140551     3.650763e+09          7889.076813   
1998                   121599     3.617070e+09         11850.394701   
2014                   151368     3.161710e+09          8885.938884   
2013                   146377     3.287065e+09          9348.631876   
2004                   144379     2.124797e+09          9932.250643   
2010                   148789     3.056312e+09          9552.091071   
2016                   137643     3.818861e+09          7079.257780   
2001                   153935     3.147821e+09         11351.712598   
1994                   131059     3.364163e+09         13138.051620   
2012                   150117     3.414637e+09          9690.506589   
2011                   150126     3.052052e+09          9654.865318   
2008                   148324     2.513047e+09          9692.425107   
2003                   149317     2.287121e+09         10176.936748   
2002                   145027     2.585502e+09         10007.324112   
2007                   146675     2.573119e+09         10146.862089   
2005                   153585     2.244703e+09         10324.548458   
2006                   145653     2.432879e+09         10115.646333   
2017                   130209     3.329380e+09          6733.883648   
1995                   132277     4.458192e+09         12784.441067   
1999                   113634     3.445345e+09         11086.829479   

             fuel_cost_per_unit_burned  fuel_cost_per_unit_delivered  \
report_year                                                            
1997                        163533.452                  1.218738e+07   
1996                         86425.275                  1.000602e+07   
2000                        326495.434                  1.352903e+06   
2009                        263584.566                  7.975923e+05   
2018                        221575.426                  4.748057e+05   
2015                        260060.984                  3.569033e+05   
1998                        140238.656                  3.178799e+05   
2014                        518002.469                  2.256952e+05   
2013                        306933.236                  2.065968e+05   
2004                        184637.699                  1.663129e+05   
2010                        331581.467                  1.158381e+05   
2016                        255484.814                  1.074344e+05   
2001                        180365.641                  8.142834e+04   
1994                        263292.115                  7.859053e+04   
2012                        321693.544                  7.581617e+04   
2011                        374175.603                  7.412059e+04   
2008                        314312.022                  7.194631e+04   
2003                        149175.726                  6.740849e+04   
2002                        121921.135                  5.735121e+04   
2007                        256114.312                  5.476283e+04   
2005                        135924.139                  5.258506e+04   
2006                        188142.032                  4.805125e+04   
2017                        244557.823                  4.587348e+04   
1995                        425661.488                  3.931506e+04   
1999                        126056.285                  2.682921e+04   

             fuel_cost_per_mmbtu  
report_year                       
1997                  3406.15391  
1996                  2933.11773  
2000      